apps/managedidentity/managedidentity.go

```diff
@@ -129,24 +169,91 @@ func New(id ID, options ...ClientOption) (Client, error) {
 	default:
 		return Client{}, fmt.Errorf("unsupported type %T", id)
 	}
+
 	client := Client{
 		miType:     id,
 		httpClient: opts.httpClient,
+		source:     source,
 	}
 
 	return client, nil
 }
 
+// Detects and returns the managed identity source available on the environment.
+func GetSource(id ID) (Source, error) {
+	identityEndpoint := os.Getenv(IdentityEndpointEnvVar)
+	identityHeader := os.Getenv(IdentityHeaderEnvVar)
+	identityServerThumbprint := os.Getenv(IdentityServerThumbprintEnvVar)
+	msiEndpoint := os.Getenv(MsiEndpointEnvVar)
+	imdsEndpoint := os.Getenv(ArcIMDSEnvVar)
+
+	if identityEndpoint != "" && identityHeader != "" {
+		if identityServerThumbprint != "" {
+			return ServiceFabric, nil
+		}
+		return AppService, nil
+	} else if msiEndpoint != "" {
+		return CloudShell, nil
+	} else if validateAzureArcEnvironment(identityEndpoint, imdsEndpoint, runtime.GOOS) {
+		return AzureArc, nil
+	}
+
+	return DefaultToIMDS, nil
+}
+
+// Acquires tokens from the configured managed identity on an azure resource.
+//
+// Resource: scopes application is requesting access to
+// Options: [WithClaims]
+func (client Client) AcquireToken(ctx context.Context, resource string, options ...AcquireTokenOption) (base.AuthResult, error) {
+	o := AcquireTokenOptions{}
+	var req *http.Request
+	var err error
+
+	for _, option := range options {
+		option(&o)
+	}
+	var tokenResponse accesstokens.TokenResponse
+	switch client.source {
+	case AzureArc:
+		req, err = createAzureArcAuthRequest(ctx, client.miType, resource, o.claims)
+		if err != nil {
+			return base.AuthResult{}, err
+		}
+
+		// need to perform preliminary request to retrieve the secret key challenge provided by the HIMDS service
+		// this is done when we get a 401 response, which will be handled by the response handler
+		tokenResponse, err = client.getTokenForRequest(ctx, req, client.handleAzureArcResponse)
+		if err != nil {
+			return base.AuthResult{}, err
+		}
+	default:
```

Comment:
I think you should treat `DefaultToImds` like this, but the rest of the cases (AppService etc.) should throw some sort of "not implemented exception".

Reply:
Right, something like
```
case DefaultToIMDS:
    // ... the current default case ...
default:
    return fmt.Errorf("unsupported source %q", client.source)
```
would be tidier